client/mainmenu/CMainMenu.cpp

```diff
@@ -371,6 +372,45 @@ void CMainMenu::openCampaignScreen(std::string name)
 	logGlobal->error("Unknown campaign set: %s", name);
 }
 
+void CMainMenu::startTutorial()
+{
+	ResourceID tutorialMap("Maps/Tutorial.tut", EResType::MAP);
+	if(!CResourceHandler::get()->existsResource(tutorialMap))
+	{
+		CInfoWindow::showInfoDialog(CGI->generaltexth->translate("core.genrltxt.742"), std::vector<std::shared_ptr<CComponent>>(), PlayerColor(1));
+		return;
+	}
+		
+	auto mapInfo = std::make_shared<CMapInfo>();
+	mapInfo->mapInit(tutorialMap.getName());
+	
+	CMainMenu::openLobby(ESelectionScreen::newGame, true, nullptr, ELoadMode::NONE);
+	
+	std::thread waitForConnectionThread([mapInfo](){
+		boost::this_thread::sleep(boost::posix_time::milliseconds(100)); //delay this thread
+		
+		//connecting to server
+		while(CSH->state != EClientState::LOBBY)
+		{
+			if(CSH->state == EClientState::CONNECTION_CANCELLED || CSH->state == EClientState::NONE)
+				return;
+			boost::this_thread::sleep(boost::posix_time::milliseconds(50));
+		}
+		
+		//start game from main thread
+		GH.dispatchMainThread([mapInfo]()
+		{
+			while(!CSH->si || mapInfo->fileURI != CSH->si->mapname)
+			{
+				CSH->setMapInfo(mapInfo);
+				boost::this_thread::sleep(boost::posix_time::milliseconds(50));
```

Comment:
Still sleeps in main thread

Reply:
Why do even need these threads / sleeps?
Normal scenario startup works just fine without all of this

Reply:
scenario startup is performed in 3 steps:
1) connect to server (show connection pop-up), separate thread waits for connection
2) set scenario by sending request to the server - main thread waits for response
3) send game start request - main thread, same as for single scenario

Reply:
Do we actually need this code which is:
- copy-pasted version of debug map loading
- introduces even more threading to our already overcomplicated threading model?

I already lost count how many threading-related bugs I have fixed recently.  Do we really need to add more code to it?

Reply:
what is bad about multithreading code? I suppose if we want tutorial with no ui freeze, than yes multithreading code is needed. As for bugs, functionality is isolated, I don't see regression risk

Reply:
And what's good in spawning dozens of threads and then attempt to synchronize all of them? What's good in copy-pasting code instead of reusing it?

Do we require some CPU heavy operations that we need to run in a separate thread?

Why not set callback that game should call when connection has been established? Why not unify debug code from which you copy-pasted this block and your new code?

> As for bugs, functionality is isolated, I don't see regression risk

Just like battle replay that broke leveling up on winning against attacking enemy?
Just like new boat types that broke existing boat logic due to uninintialized variables?
Just like configurable buildings in towns that broke visiting of towns by heroes, build all cheat and music playback after recruiting new hero?

Reply:
> If you want to wait for local server, it should be almost instaneous (unless there are other problems in the system). Then don't delay it by adding 100 ms sleep at the beginning.

Far from it. 200-300 ms is already "fast". Might take up to 5 seconds on low-end systems, especially when mods are used.

> I also don't see the benefit for unlocking user's GUI while he already decided to launch a tutorial.

At the very least software mouse pointer will freeze and OS will interpret vcmi as hanging app while sleep is ongoing. GUI needs to be unlocked. Input on the other hand might be ignored/blocked if needed.

Reply:
> Starting local server takes couple seconds, but 100ms is needed to ensure that server launching thread if started before waiting for connection thread

Sleep is not a thread synchronization mechanism. If you need to wait for another thread use mutex / conditional variable / atomic / callback / etc, depending on what's appropriate

> I see one new thread, which waits for connection to server in order to not block popup "connecting to server" window, why we are talking about dozens?

One thread here, one thread there and we already have 13 different threads just in game client (not counting vcmi_lib which also spawns own threads as well as threads from SDL which also interact with our code). Each of these threads complicate code and make tracking notoriously hard multithreading issues even more complex.

> This part is toxic and counter-productive. I don't know how am I supposed to react

Problem is, unless you plan to stick around for a longer period of time, any bugs in this code are for me to resolve. Sure, this code is isolated and won't be called unless you click "Tutorial" button but it is not like we can ask players not to click this button in a wrong way.

So please, keep multithreading to minimum to reduce already painful debugging efforts.

Reply:
How should I contact you then?

I've tried pinging you on some iphone issue. No reply until I got reply from kambala instead. I've tried commenting on your PR with discovered regression. Also no reply. We also had issues since May that in the end turned out to be regression from combat replay PR.

Reply:
nothing special, both GitHub and slack work well usually, but we are all humans so I can miss something or be late with response, certainly no intent to ignore. But even in that case follow up is welcomed